api/document/tree.py

```diff
@@ -125,45 +217,51 @@ def nested_set_renumber(self, left=1, bulk_create=True):
         if bulk_create:
             self._bulk_create()
 
-    def _bulk_create(self):
+    def _bulk_create(self) -> None:
         DocNode.objects.bulk_create(node.model for node in self.walk())
 
-    def next_sort_order(self):
+    def next_sort_order(self) -> int:
         return self.tree.out_degree(self.identifier)
 
-    def parent(self):
+    def parent(self) -> Optional['DocCursor']:
         predecessors = list(self.tree.predecessors(self.identifier))
         if predecessors:
             return type(self)(self.tree, predecessors[0])
+        return None
 
-    def ancestors(self, filter_fn=None):
+    def ancestors(self, filter_fn: Callable[['DocCursor'], bool]=None) \
+            -> Iterator['DocCursor']:
         parent = self.parent()
         if parent and (not filter_fn or filter_fn(parent)):
             yield parent
         if parent:
             yield from parent.ancestors(filter_fn)
 
-    def add_models(self, models):
+    def add_models(self, models: Iterator[DocNode]) -> 'DocCursor':
         """Convert a (linear) list of DocNodes into a tree-aware version.
         We assume that the models are already sorted."""
         parent = self
         for child in models:
             # not a child of this parent; move cursor up
             while child.left > parent.right:
-                parent = parent.parent()
+                parents_parent = parent.parent()
+                if parents_parent is None:
+                    raise AssertionError("Parent's parent must exist!")
```

Comment:
Looks like mypy found an edge case here! I'm type narrowing and raising an error if `parent.parent()` is `None`, but I'm not sure what the best assertion message to raise is, since I'm still not very familiar with the nested set algorithm--any suggestions?

Reply:
Awesome. I think this case could arise if either:
1. The DocNodes weren't given in the right sort order
2. The DocNodes don't have sensible left/right values (they didn't go through our nested set accounting)

So maybe something along the lines of:
"Couldn't convert DocNodes into a tree. Perhaps they weren't sorted or there was data corruption?"